node/lib/util/stitch_util.js

```diff
@@ -0,0 +1,637 @@
+            }
+        } else if (null !== newSha && null !== adjustPath(name)) {
+            synthetics.push(name);
+            yield stitchSub(name, newSha);
+        }
+    }
+
+    // If any kept submodules were added or removed, rewrite the modules
+    // file.
+
+    if (updateModules) {
+        const newUrls =
+               yield SubmoduleConfigUtil.getSubmodulesFromCommit(repo, commit);
+        const content = yield exports.computeModulesFile(repo,newUrls,
+                                                         keepAsSubmodule,
+                                                         adjustPath);
+        changes[SubmoduleConfigUtil.modulesFileName] = content;
+    }
+
+    // Adjust paths using the provided `adjustPath` function.
+
+    const mappedChanges = {};
+
+    for (let name in changes) {
+        const mapped = adjustPath(name);
+        if (null !== mapped) {
+            mappedChanges[mapped] = changes[name];
+        }
+    }
+
+    const newTree = yield TreeUtil.writeTree(repo, parentTree, mappedChanges);
+
+    let newCommit = null;
+    let mappedSha = null;
+    if (skipEmpty && exports.isEmpty(newTree, parentTree)) {
+        if (0 !== parents.length) {
+            mappedSha = parents[0].id().tostrS();
+        }
+    } else {
+        const newCommitId = yield NodeGit.Commit.create(
+                                                      repo,
+                                                      null,
+                                                      commit.author(),
+                                                      commit.committer(),
+                                                      commit.messageEncoding(),
+                                                      commitMessage,
+                                                      newTree,
+                                                      parents.length,
+                                                      parents);
+        newCommit = yield repo.getCommit(newCommitId);
+        mappedSha = newCommit.id().tostrS();
+    }
+    if (null !== mappedSha) {
+        const convertedRef = exports.convertedRefName(commit.id().tostrS());
+        yield NodeGit.Reference.create(repo,
+                                       convertedRef,
+                                       mappedSha,
+                                       1,
+                                       "stitched a ref");
+    }
+    return newCommit;
+});
+
+/**
+ * In the specified `repo`, perform the specified `subFetches`.  Use the
+ * specified `url` to resolve relative submodule urls.  Each entry in the
+ * `subFetches` array is an object containing the fields:
+ * 
+ * - url -- submodule configured url
+ * - sha -- submodule sha
+ * - metaSha -- sha it was introcued on
+ *
+ * @param {NodeGit.Repository}  repo
+ * @param {String}              url
+ * @param {[Object]}            subFetches
+ */
+exports.fetchSubCommits = co.wrap(function *(repo, url, subFetches) {
```

Comment:
This is, in general, super inefficient.  In 99% of cases, submodule commits will be fast-forward, so if we fetch the most-recent one we can then check to see if we have each remaining one before trying to fetch it.  And if we do have it, we don't need to write a synthetic ref for it, since that means that the synthetic ref written for the first thing we fetched will contain it.

Reply:
`GitUtil.fetchSha` checks to see if the commit already exists before attempting to fetch it:

```JavaScript
/**
 * Fetch the specified `sha` from the specified `url` into the specified
 * `repo`, if it does not already exist in `repo`.
 *
 * @async
 * @param {NodeGit.Repository} repo
 * @param {String}             url
 * @param {String}             sha
 */
exports.fetchSha  = co.wrap(function *(repo, url, sha) {
```

Reply:
Can we make this run in most-recent-first order?  And do we want to avoid writing synthetic refs for shas that we didn't have to fetch?

Reply:
I do fetch them in that order (otherwise it would take months or years to run), but I'm making the synthetic refs whether a fetch is needed or not. I can change this to not make the ref unless a fetch happens (might need a tweak in the api for `GitUtil.fetchSha` to return a bool).